About docs (#2)
broccolijs/broccolijs.github.io: src/docs/about.md

---
title: What is Broccoli?
---

> Broccoli is a simple Javascript build tool, that allows you to configure your build pipeline in Javascript (like you're

Reviewer: JavaSciprt

> you're already used to writing), and handles setting up the filesystem state for each transformation that's 

Reviewer 2: I would not necessarily assume that readers understand the distinction between an asset pipeline and a module bundler.

You could say something like

> Broccoli is a build tool, that through allowing developers to compose modular plugins to form an asset pipeline, quickly and reliably combines and transforms source files into deployable static assets.

${toc}

## What is a build tool?

A build tool's job is to take input files (your javascript, css, html, etc), to process them and output the result of
those transformations, usually into some distributable form. Typically this will involve things like javascript 
transformations to allow you to write newer syntax that will work in a browser, to use things like Sass for your CSS,
etc.

Broccoli.js is different to other build tools. You may be used to tools like [Grunt](https://gruntjs.com/) (a task
runner), [Gulp](https://gulpjs.com/) (streams and pipes) or [Webpack](https://webpack.js.org/) (a module bundler),
these all operate on different levels to Broccoli.

Broccoli works at the file-system level, it provides a Javascript API to wrap other node tools like

> [Babel](https://babeljs.io/), [Rollup](https://rollupjs.org/) or [Node-Sass](https://github.com/sass/node-sass), that

Reviewer: webpack, browserify, and many more

operate on a set of input files, and emit output files, that is all. Broccoli has no knowledge of the contents of 
your files, only input and output directories.

## Thinking in Broccoli

> Broccoli at its heart is very simple. It's really just a set of connected file transformations, performed by plugins,

Reviewer: Maybe something like:

It is really just a set of directories, connected together by a plugin system. This plugin systems allow behavior between two directories to be described. Broccoli then ensures plugins are called at the appropriate time, to ensure the build is succesfull.

> Broccoli at its heart is very simple. It is really just a set of directories, connected together by a plugin

Reviewer 2: ```suggestion
Broccoli consists of a set of directories, connected together by plugins, which describe how files are moved or transformed at each step of the build process.
```

> system. This plugin systems allow behavior between input and output directories to be described. Broccoli then

Reviewer 2: ```suggestion
Broccoli
```

> Broccoli handles managing the file-system state, and simply takes input files and passes them as inputs to plugins, 

Reviewer: broccoli doesn't really care about files, it's simply connects directories and behavior together to achieve what the user wants. Custom behavior is provided by plugins, which are  responsible to read from their input directories, and write to their single output directory.

> ensures plugins are called at the appropriate time, to ensure the build is successful, and writes the result to

Reviewer 2: ```suggestion
ensures plugins are called in the prescribed order, and writes the result to
```

and takes the output from each plugin and passes it as an input to the next one. The order in which these operations 
happen is determined by the build file, a `Brocfile.js` that you're going to learn how to create.

> You can think of Broccoli much like a simple programming language, where the output of a function can be passed as the

Reviewer: maybe `s/broccoli/broccoli-plugins`

plugins, creates an output directory for the plugin to write to, and passes that output directory it as an input

> to the next plugin.

Reviewer 2: You're taking a lot of time to try and describe something that might be easier to show via a diagram

```
app                                       dist
 └─ src						         	   └─ prod.build-20190206.js
     ├─ index.ts      -->  MyPlugin -->
     └─ other.ts

```

E.g.:

```js
module.exports = merge([
    uglify(
        babel('src')
    ),
    autoprefixer(
        sass('src/styles', 'site.scss')
    )
]);
```

> In the above, working from inner to outer, a `src` directory is passed to the `babel()` plugin (which will convert our

Reviewer: Rather than saying "from inner to outer" might it be clearer to have a number list of steps which sequentializes the pipeline in question? I at-least find "inner to outer" to be hard to follow.

new ES6 syntax into ES5 that runs in the browser), the output of that is passed to the `uglify()` plugin which will
minify our JS into a smaller format. The output of `uglify()` will in turn be passed as 1 input to `merge()`.

Additionally, a `src/styles` directory and the input file `site.scss` is passed to `sass()` which will convert your
`.scss` files into `.css` files, its output (your css) is passed as an input to `autoprefixer()` which will add vendor 
prefixes (like `-ms` or `-webkit`) to attributes, which is then in turn passed into the `merge()` plugin.

Merge will copy the contents of each of its inputs into its output directory. Thus, it merges our uglified Javascript
and out vendor prefixed css. This then becomes our final output and is what is written to our target (destination) 
directory.

This should all be fairly familiar to you if you've ever written Javascript (or any programming language for that
matter) before, it's just inputs and output.

## Thinking in Broccoli

There are 3 main concepts to get your head around when using Broccoli:

* [Plugins](#plugins)
* [Nodes](#nodes)
* [Trees](#trees)

### Plugins

> Plugins are what a build pipeline developer will interact with most. Plugins are what do the actual work of transforming

Reviewer: We spoke about this in todays call, I will review again once you had time to iterate on this section.

vendor  prefixes (like `-ms` or `-webkit`) to attributes, which is then in turn passed into the `merge()` plugin.

The `merge()` plugin will copy the contents of each of its inputs into its output directory. Thus, it merges our 
uglified JavaScript and out vendor prefixed css. This then becomes our final output and is what is written to 
our target (destination) directory.

This should all be fairly familiar to you if you've ever written JavaScript (or any programming language for that
matter) before, it's just inputs and output.

## Thinking in Broccoli

There are 2 main concepts to get your head around when using Broccoli:

* [Plugins](#plugins)
* [Trees](#trees)

### Plugins

Plugins are what a build pipeline developer will interact with most. Plugins are what do the actual work of

> transforming input files into output files. The API of a plugin is pretty simple, all that's required is creating

Reviewer 2: you may want to be careful of input/output as terminology, since they could be taken to mean pipeline imports or plugin imports.

```suggestion
transforming files at each step of build process. The API of a plugin is pretty simple, all that's required is creating
```

a class that extends the [broccoli-plugin](https://github.com/broccolijs/broccoli-plugin) base class, and
implementing a `build()` method, that performs some work or returns a promise.

```js
const Plugin = require('broccoli-plugin');

class MyPlugin extends Plugin
{
  build() {
    // A plugin can receive single or multiple inputs nodes/directories
    // The files are available in this.inputPaths[0]/this.inputPaths[1]...
    // You can do anything with these files that you can do in node
    // The output of your plugin must write to this.outputPath
  }
}
```

> A broccoli-plugin has only 1 purpose to transform the files from `this.inputPaths` directories to their output

Reviewer 2: ```suggestion
A broccoli-plugin has only one purpose: to transform the files from `this.inputPaths` directories to their output
```

There is a special case where a `string` is passed as an input to a plugin. When parsing your build pipeline, 
Broccoli will automatically convert a string input into a
[source plugin](https://github.com/broccolijs/broccoli-source). This plugin basically connects its input directory
directly to its output directory, and also allows Broccoli to `watch` and be notified when files within the input
directory change and trigger a rebuild. You can also manually create an `unwatched` directory from a string by
using [UnwatchedDir](https://github.com/broccolijs/broccoli-source#new-unwatcheddirdirectorypath-options).

### Trees

Working from the target directory (the final output) back up to the source directories resembles a tree. Think 
of a piece of Broccoli and you should have a mental model of what a build pipeline looks like.

Broccoli constructs this tree of connected nodes in memory as it parses the `Brocfile`, and sets up the filesystem 
state for each node, creating an `outputPath` for each node to write to.

You may often encounter the term "tree" when reading plugin READMEs or in tutorials, just remember a tree is a

> connected set of plugins/nodes.

Reviewer 2: One of the important concepts to get across is that trees can be 

- a string
- something a plugin emits

## Building

> Broccoli build pipelines are defined using a `Brocfile.js` file in the root of the project. This `js` file sets up

Reviewer 2: > Broccoli build pipelines are defined using a `Brocfile.js` file in the root of the project.
feels like you've said this a few times already

> the build graph, connecting source directories via plugins, and exports a single node whose output will be written

Reviewer 2: Is it important that consumers of broccoli know about a "build graph"? This sounds like it might be an implementation detail

---
title: What is Broccoli?
---

> Broccoli is a simple JavaScript build tool, that allows you to configure your build pipeline in Javascript (like

Reviewer: `Javascript` -> `JavaScript`

you're already used to writing), and handles setting up the filesystem state for each transformation that's 
going to happen.

${toc}

## What is a build tool?

A build tool's job is to take input files (your javascript, css, html, etc), to process them and output the result
of those transformations, usually into some distributable form. Typically this will involve things like JavaScript 
transformations to allow you to write newer syntax that will work in a browser, to use a CSS pre-processor like 
Sass for your CSS, etc.

Broccoli.js is different to other build tools. You may be used to tools like [Grunt](https://gruntjs.com/) (a task
runner), [Gulp](https://gulpjs.com/) (streams and pipes) or [Webpack](https://webpack.js.org/) (a module bundler),
these all operate on different levels to Broccoli.

Broccoli works at the file-system level, it provides a Javascript API to wrap other node tools like
[Babel](https://babeljs.io/), [Rollup](https://rollupjs.org/) or [Node-Sass](https://github.com/sass/node-sass), 

> [Webpack](https://webpack.js.org/), [Browserify](http://browserify.org/), and many more. These tools operate on a

Author: Correct, suggestions welcome

our target (destination) directory.

This should all be fairly familiar to you if you've ever written Javascript (or any programming language for that
matter) before, it's just inputs and output.

## Thinking in Broccoli

There are 3 main concepts to get your head around when using Broccoli:

* [Plugins](#plugins)
* [Nodes](#nodes)
* [Trees](#trees)

### Plugins

Plugins are what a build pipeline developer will interact with most. Plugins are what do the actual work of
transforming input files into output files. The API of a plugin is pretty simple, all that's required is creating
a class that extends the [broccoli-plugin](https://github.com/broccolijs/broccoli-plugin) base class, and
implementing a `build()` method, that performs some work or returns a promise.

```js
const Plugin = require('broccoli-plugin');

class MyPlugin extends Plugin
{
  build() {
    // A plugin can receive single or multiple inputs nodes/directories
    // The files are available in this.inputPaths[0]/this.inputPaths[1]...
    // You can do anything with these files that you can do in node
    // The output of your plugin must write to this.outputPath
  }
}
```

> Basic plugins are super simple, anything you can do in node, you can do in the `build()` method. A plugin can

Reviewer: I think we should avoid saying phrases like `super simple`. Rather lets show how they are "super simple". For example, "a broccoli-plugin has only 1 purpose to transform the files from this.inputPaths directories to their appropriate output in this.outputPath directory when its build function is invoked."

receive one or multiple inputs, and these are available in the `this.inputPaths` array in the order they are 
provided. `this.inputPaths` contains paths to directories, that are the `outputPath` of previous plugins. Each 
`inputPath` contains files that you can manipulate and write to `this.outputPath`. Broccoli will handle the 
state of these directories and passing them between plugins.

There is a special case where a `string` is passed as an input to a plugin. When parsing your build pipeline, 
Broccoli will automatically convert a string input into a
[source plugin](https://github.com/broccolijs/broccoli-source). This plugin basically connects its input directory
directly to its output directory, and optionally allows Broccoli to be notified when files within the input 
directory change and trigger a rebuild.

> ### Nodes

Reviewer: Do you think we can get away without talking about nodes? I think for the most part, just thinking about broccoli-plugins being connected by broccoli in a graph is sufficient. Yes that implies broccoli-plugins are nodes, but I worry raising nodes and pointing to graph-theory provides less value than cost.

Author: Yeah I think that's fair.
